sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala

```diff
@@ -606,9 +606,17 @@ abstract class HadoopFsRelation private[sql](
         // we need to cast into the data type that user specified.
         def castPartitionValuesToUserSchema(row: InternalRow) = {
           InternalRow((0 until row.numFields).map { i =>
-            Cast(
-              Literal.create(row.getString(i), StringType),
-              userProvidedSchema.fields(i).dataType).eval()
+            row.isNullAt(i) match {
```

Comment:
@davies Thanks for reviewing the change. In cases when we know in advance that schema does not allow nulls , sometimes we can skip this null check. By moving it to internalRow , would we loose the opportunity to optimize ? Pl. let me know.